src/mpid/ch4/shm/posix/posix_coll_nb_release_gather.h

```diff
@@ -0,0 +1,132 @@
+/*
+ *
+ *  Copyright (C) by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ */
+
+#ifndef POSIX_COLL_NB_RELEASE_GATHER_H_INCLUDED
+#define POSIX_COLL_NB_RELEASE_GATHER_H_INCLUDED
+
+#include "mpiimpl.h"
+#include "algo_common.h"
+#include "nb_release_gather.h"
+
+/* Intra-node Ibcast is implemented as a release step followed by gather step in release_gather
+ * framework. The actual data movement happens in release step. Gather step makes sure that
+ * the shared bcast buffer can be reused for next bcast call. Release gather framework has
+ * multitple cells in bcast buffer, so that the copying in next cell can be overlapped with
+ * copying out of previous cells (pipelining).
+ */
+MPL_STATIC_INLINE_PREFIX int MPIDI_POSIX_ibcast_release_gather(void *buffer, int count,
+                                                               MPI_Datatype datatype, int root,
+                                                               MPIR_Comm * comm_ptr,
+                                                               MPIR_TSP_sched_t * sched)
+{
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_POSIX_MPI_IBCAST_RELEASE_GATHER);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_POSIX_MPI_IBCAST_RELEASE_GATHER);
+
+    int mpi_errno = MPI_SUCCESS;
+
+    /* If there is only one process or no data, return */
+    if (count == 0 || (MPIR_Comm_size(comm_ptr) == 1)) {
+        goto fn_exit;
+    }
+
+    MPIDI_POSIX_COMM(comm_ptr, nb_release_gather).num_collective_calls++;
+    if (MPIDI_POSIX_COMM(comm_ptr, nb_release_gather).num_collective_calls <
+        MPIR_CVAR_POSIX_NUM_NB_COLLS_THRESHOLD) {
+        /* Fallback to pt2pt algorithms if the total number of release_gather collective calls is
+         * less than the specified threshold */
+        goto fallback;
+    }
+
+    /* Lazy initialization of release_gather specific struct */
+    mpi_errno =
+        MPIDI_POSIX_nb_release_gather_comm_init(comm_ptr, MPIDI_POSIX_RELEASE_GATHER_OPCODE_BCAST);
+    MPII_COLLECTIVE_FALLBACK_CHECK(MPIR_Comm_rank(comm_ptr), !mpi_errno, mpi_errno,
+                                   "release_gather ibcast cannot create more shared memory. Falling back to pt2pt algorithms.\n");
```

Comment:
I'd prefer to see the details of these comm_init functions moved into this commit, rather than part of the framework commit. That way we can more easily evaluate the algorithm addition altogether.

Reply:
@raffenet - Do you want the framework and collective algorithms to be in the same commit. The algorithms cannot be evaluated without the non-blocking compositions. Everything in this PR can be tested after adding the compositions.